web-ng/src/app/components/layer-dialog/layer-dialog.component.ts

```diff
@@ -14,24 +14,86 @@
  * limitations under the License.
  */
 
-import { Component, Inject, OnInit } from '@angular/core';
+import { Component, Inject, OnDestroy } from '@angular/core';
+import { Observable } from 'rxjs';
+import { ProjectService } from '../../services/project/project.service';
 import { MatDialogRef, MAT_DIALOG_DATA } from '@angular/material/dialog';
+import { Project } from '../../shared/models/project.model';
+import { Layer } from '../../shared/models/layer.model';
+import { Subscription } from 'rxjs';
+import { DataStoreService } from '../../services/data-store/data-store.service';
+import { Router } from '@angular/router';
 
 @Component({
   selector: 'app-layer-dialog',
   templateUrl: './layer-dialog.component.html',
   styleUrls: ['./layer-dialog.component.css'],
 })
-export class LayerDialogComponent {
+export class LayerDialogComponent implements OnDestroy {
   layerId: string;
-
+  layer?: Layer;
+  projectId?: string;
+  activeProject$: Observable<Project>;
+  subscription: Subscription = new Subscription();
   constructor(
     // tslint:disable-next-line:no-any
     @Inject(MAT_DIALOG_DATA) data: any,
-    private dialogRef: MatDialogRef<LayerDialogComponent>
+    private dialogRef: MatDialogRef<LayerDialogComponent>,
+    private projectService: ProjectService,
+    private dataStoreService: DataStoreService,
+    private router: Router
   ) {
-    this.layerId = data.layerId!;
     // Disable closing on clicks outside of dialog.
     dialogRef.disableClose = true;
+    this.layerId = data.layerId;
+    this.activeProject$ = this.projectService.getActiveProject$();
+    this.subscription.add(
+      this.activeProject$.subscribe(project => {
+        this.onProjectLoaded(project);
+      })
+    );
+  }
+
+  onProjectLoaded(project: Project) {
+    if (this.layerId === ':new') {
+      this.layerId = this.dataStoreService.generateId();
+      this.layer = {
+        id: this.layerId,
+      };
+    } else {
+      this.layer = project.layers.get(this.layerId);
+    }
+    if (!this.layer) {
+      throw Error('No layer exists');
+    }
+    this.projectId = project.id;
+  }
+
+  onSave() {
+    // TODO: Wait for project to load before showing dialog.
+    if (!this.projectId) {
+      throw Error('Project not yet loaded');
+    }
+    const layer = {
+      id: this.layerId,
+    };
+    try {
+      // TODO: Inform user layer was saved
+      this.dataStoreService
+        .updateProjectLayer(this.projectId, layer)
+        .then(() => this.onClose());
+    } catch (err) {
```

Comment:
You'll need to catch the error in .catch(). Not sure if JavaScript throw/catch will do here.

Reply:
I tried passing bad data value for updated and it is getting caught in the catch block that I created :). Also when I passed bad data value for update it didn't got caught in the .catch() block

Reply:
That's really odd! Can you paste the stack trace here so we can figure out where the exception is being thrown?

Reply:
See also Firestore API docs where they show how to catch errors in the returned Promise:
https://firebase.google.com/docs/firestore/manage-data/add-data#update-data